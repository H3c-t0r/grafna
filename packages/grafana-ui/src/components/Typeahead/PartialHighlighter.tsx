import React from 'react';
import { HighlightPart } from '../../types';

interface Props {
  text: string;
  highlightParts: HighlightPart[];
  highlightClassName: string;
}

export const PartialHighlighter: React.FC<Props> = (props: Props) => {
  let { highlightParts, text, highlightClassName } = props;

  let children = [];

  const firstRange = highlightParts[0];

  if (firstRange.start !== 0) {
    children.push(<span key={children.length}>{text.substring(0, firstRange.start)}</span>);
  }

  for (let i = 0; i < highlightParts.length - 1; i++) {
    let range = highlightParts[i],
      nextRange = highlightParts[i + 1];
    children.push(
      <mark key={children.length} className={highlightClassName}>
        {text.substring(range.start, range.end + 1)}
      </mark>
    );
    children.push(<span key={children.length}>{text.substring(range.end + 1, nextRange.start)}</span>);
  }

  let lastRange = highlightParts[highlightParts.length - 1];

  children.push(
    <mark key={children.length} className={highlightClassName}>
      {text.substring(lastRange.start, lastRange.end + 1)}
    </mark>
  );

  if (lastRange.end + 1 < text.length) {
    children.push(<span key={children.length}>{text.substring(lastRange.end + 1, text.length)}</span>);
  }

  return <div>{children}</div>;
};
