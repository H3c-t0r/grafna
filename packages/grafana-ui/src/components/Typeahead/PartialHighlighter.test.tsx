import React from 'react';
import { shallow } from 'enzyme';
import { PartialHighlighter } from './PartialHighlighter';

describe('PartialHighlighter component', () => {
  it('should highlight inner parts correctly', () => {
    const wrapper = shallow(
      <PartialHighlighter
        text="Lorem ipsum dolor sit amet"
        highlightClassName="highlight"
        highlightParts={[
          { start: 6, end: 10 },
          { start: 18, end: 20 },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should highlight outer parts correctly', () => {
    const wrapper = shallow(
      <PartialHighlighter
        text="Lorem ipsum dolor sit amet"
        highlightClassName="highlight"
        highlightParts={[
          { start: 0, end: 4 },
          { start: 22, end: 25 },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
