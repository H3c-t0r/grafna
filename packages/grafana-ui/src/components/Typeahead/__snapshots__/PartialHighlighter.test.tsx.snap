// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartialHighlighter component should highlight inner parts correctly 1`] = `
<div>
  <span
    key="0"
  >
    Lorem 
  </span>
  <mark
    className="highlight"
    key="1"
  >
    ipsum
  </mark>
  <span
    key="2"
  >
     dolor 
  </span>
  <mark
    className="highlight"
    key="3"
  >
    sit
  </mark>
  <span
    key="4"
  >
     amet
  </span>
</div>
`;

exports[`PartialHighlighter component should highlight outer parts correctly 1`] = `
<div>
  <mark
    className="highlight"
    key="0"
  >
    Lorem
  </mark>
  <span
    key="1"
  >
     ipsum dolor sit 
  </span>
  <mark
    className="highlight"
    key="2"
  >
    amet
  </mark>
</div>
`;
